File: medium/count-battle-ships-dfs.js
class solution {
    constructor(board) {
        // TODO: find a way to calculate different directions
        // this.directions = new Array([]);
        this.board = new Array([]);
        this.m = board.length;
        this.n = board[0].length;
        this.visited = new Array(this.m[this.n]);
        this.count = 0;
        this.directions = new Array([]);

        this.nextRow = 0;
        this.nextCol = 0;
    }

    
    countBattleshipsDFS(board = new Array([])) {
        for (let row = 0; row < this.m.length; row++) {
            for (let col = 0; col < this.n.length; col++) {
                if (board[row][col] === 'X' && !this.visited[row][col]) {
                    this.count++;
                    this.dfs(board, row, col, this.visited)
                }
            }
        }
    }

    dfs(board = new Array([]), i, j, visited = new Array([])) {
        let directions = new Array([]);
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || board[i][j] || visited[i][j]) {
            return;
        }

        visited[i][j] = true;
        
        for (let d in directions) {
            this.nextRow = i + d[0];
            this.nextCol = j + d[1];

            this.dfs(board, nextRow, nextCol, visited);
        }
    }

    getDirection(ship) {
        if (ship !== null) {
            
        }
    }
}

let solved = new solution()
solved.countBattleshipsDFS([]);

File: Java/medium/count-battleships-dfs.java
class Solution {
    private static final int[][] directions = new int[][] {
        {0. 1},
        {1, 0}
    };

    public int countBattleShipsDFS(char[][] board) {
        if (board == null || board.length == 0 || board[0].length == 0) {
            return 0
        }

        int m = board.length;
        int n = board[0].length;

        boolean[][] visited = new boolean[m][n];
        int count = 0;

        for (int row = 0; row < m; row++) {
            for (int col = 0; col < n; col++) {
                if (board[row][col] == 'x' && !visited[row][col]) {
                    count++;
                    this.dfs(board, row, col, visited);
                }
            }
        }

        return count;
    }

    private void dfs(char[][] board, int i, int j, boolean[][] visited) {
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || board[i][j] || visited[i][j]) {
            return;
        }

        visited[i][j] = true;

        for (int[] d : directions) {
            int nextRow = i + d[0];
            int nextCol = j + d[1];

            dfs(board, nextRow, nextCol, visited);
        }
    }
}